File: core/src/core/usecases/documents/requisites.py
"""Реквизиты сторон: единый набор полей для своей компании и для контрагента.

Ключи здесь короткие (``inn``, ``bic``), а в шаблоне те же реквизиты живут с
префиксом стороны (``seller_inn``, ``client_inn``) — по префиксу сценарий
и понимает, откуда подставлять значение.
"""

from __future__ import annotations

from collections.abc import Mapping

from core.domain.documents import FieldSpec, FieldType, FieldValue, validate_fields

SELLER_PREFIX = "seller_"
CLIENT_PREFIX = "client_"

REQUISITE_FIELDS: tuple[FieldSpec, ...] = (
    FieldSpec("name", "Название", FieldType.TEXT, required=True),
    FieldSpec("inn", "ИНН", FieldType.INN, required=False),
    FieldSpec("kpp", "КПП", FieldType.KPP, required=False),
    FieldSpec("ogrn", "ОГРН", FieldType.OGRN, required=False),
    FieldSpec("address", "Адрес", FieldType.ADDRESS, required=False),
    FieldSpec("director", "Подписант", FieldType.NAME, required=False),
    FieldSpec("bank", "Банк", FieldType.TEXT, required=False),
    FieldSpec("bic", "БИК", FieldType.BIC, required=False),
    FieldSpec("account", "Расчётный счёт", FieldType.ACCOUNT, required=False),
    FieldSpec("phone", "Телефон", FieldType.PHONE, required=False),
    FieldSpec("email", "Почта", FieldType.EMAIL, required=False),
)


def validate_requisites(raw: Mapping[str, str]) -> tuple[dict[str, str], list[str]]:
    """Проверить карточку теми же правилами, что и поля документа."""
    result = validate_fields(REQUISITE_FIELDS, {k: FieldValue(v) for k, v in raw.items()})
    values = {key: value.value for key, value in result.values.items()}
    return values, [error.message for error in result.errors]

File: web/src/pages/profile/requisites.ts
// Реквизиты стороны — одни и те же для своей организации и для клиента
// (core.usecases.documents.requisites). Разделы — как в макете профиля:
// юрданные, банк и подписант, контакты.
import type { FieldType } from '@/api/client';

export interface RequisiteSpec {
  key: string;
  label: string;
  type: FieldType;
  // Пример значения — подсказкой под полем.
  example?: string;
}

export const REQUISITE_GROUPS: Array<{ title: string; fields: RequisiteSpec[] }> = [
  {
    title: 'Юрданные',
    fields: [
      { key: 'name', label: 'Название', type: 'text', example: 'ООО «Ромашка»' },
      { key: 'inn', label: 'ИНН', type: 'inn' },
      { key: 'kpp', label: 'КПП', type: 'kpp' },
      { key: 'ogrn', label: 'ОГРН или ОГРНИП', type: 'ogrn' },
      { key: 'address', label: 'Юридический адрес', type: 'address' },
    ],
  },
  {
    title: 'Банк и подписант',
    fields: [
      { key: 'bank', label: 'Банк', type: 'text', example: 'ПАО Сбербанк' },
      { key: 'bic', label: 'БИК', type: 'bic' },
      { key: 'account', label: 'Расчётный счёт', type: 'account' },
      { key: 'director', label: 'Подписант', type: 'name', example: 'Иванов Иван Иванович' },
    ],
  },
  {
    title: 'Контакты',
    fields: [
      { key: 'phone', label: 'Телефон', type: 'phone', example: '+7 900 000-00-00' },
      { key: 'email', label: 'Почта', type: 'email', example: 'mail@company.ru' },
    ],
  },
];

export type Requisites = Record<string, string>;

// Сервер склеивает ошибки реквизитов в одну строку через «; »: разносим их по
// полям по названию в кавычках, нераспознанное остаётся общей ошибкой.
export function splitRequisiteErrors(message: string): {
  byField: Record<string, string>;
  rest: string[];
} {
  const labelToKey = new Map<string, string>([
    ['Название', 'name'],
    ['ИНН', 'inn'],
    ['КПП', 'kpp'],
    ['ОГРН', 'ogrn'],
    ['Адрес', 'address'],
    ['Подписант', 'director'],
    ['Банк', 'bank'],
    ['БИК', 'bic'],
    ['Расчётный счёт', 'account'],
    ['Телефон', 'phone'],
    ['Почта', 'email'],
  ]);
  const byField: Record<string, string> = {};
  const rest: string[] = [];
  for (const part of message.split(/;\s*/).filter(Boolean)) {
    const match = /^«([^»]+)»:\s*(.+)$/.exec(part);
    const key = match ? labelToKey.get(match[1] ?? '') : undefined;
    if (key && match?.[2]) byField[key] = match[2];
    else rest.push(part);
  }
  return { byField, rest };
}
